src/main/java/beam/analysis/plots/RealizedModeStats.java

```diff
@@ -0,0 +1,309 @@
+        GraphUtils.saveJFreeChartAsPNG(chart, graphImageFile, GraphsStatsAgentSimEventsListener.GRAPH_WIDTH, GraphsStatsAgentSimEventsListener.GRAPH_HEIGHT);
+    }
+
+    //This is used for removing columns if all entries is 0
+    private Set<String> getModesChosen() {
+
+        Set<String> modes = new TreeSet<>();
+        Map<String, Integer> modeCountBucket = new HashMap<>();
+        hourModeFrequency.keySet().stream().forEach(hour -> hourModeFrequency.get(hour).keySet().stream().
+                forEach(mode -> {
+                    Integer count = modeCountBucket.get(mode);
+                    Map<String, Integer> modeFrequency = hourModeFrequency.get(hour);
+                    if (modeFrequency != null) {
+                        if (count != null) {
+                            modeCountBucket.put(mode, count + modeFrequency.get(mode));
+                        } else {
+                            modeCountBucket.put(mode, modeFrequency.get(mode));
+                        }
+                    }
+                }));
+
+        modeCountBucket.keySet().stream().forEach(mode -> {
+            if (modeCountBucket.get(mode) != null && modeCountBucket.get(mode) != 0) {
+                modes.add(mode);
+            }
+        });
+        return modes;
+    }
+
+    private void writeToCSV(IterationEndsEvent event) {
+
+        String csvFileName = GraphsStatsAgentSimEventsListener.CONTROLLER_IO.getIterationFilename(event.getIteration(), fileName + ".csv");
+
+        try (final BufferedWriter out = new BufferedWriter(new FileWriter(new File(csvFileName)))) {
+
+            Set<String> modes = getModesChosen();
+
+            String heading = modes.stream().reduce((x, y) -> x + "," + y).get();
+            out.write("hours," + heading);
+            out.newLine();
+
+            int max = hourModeFrequency.keySet().stream().mapToInt(x -> x).max().getAsInt();
+
+            for (int hour = 0; hour <= max; hour++) {
+                Map<String, Integer> modeCount = hourModeFrequency.get(hour);
+                StringBuilder builder = new StringBuilder(hour + 1 + "");
+                if (modeCount != null) {
+                    for (String mode : modes) {
+                        if (modeCount.get(mode) != null) {
+                            builder.append("," + modeCount.get(mode));
+                        } else {
+                            builder.append(",0");
+                        }
+                    }
+                } else {
+                    for (String mode : modes) {
+                        builder.append(",0");
+                    }
+                }
+                out.write(builder.toString());
+                out.newLine();
+            }
+            out.flush();
+        } catch (IOException e) {
+            log.error("CSV generation failed.", e);
+        }
+    }
+
+    class ModePerson {
```

Comment:
❗️Consider making this class static otherwise every instance will have a reference to the enclosing class. This increase the memory usage and can lead to memory leaks.